### ISSUE TITLE
A Large-Scale Study of Programming Languages and Code Quality in Github | October 2017

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6413]
A Large-Scale Study of Programming Languages and Code Quality in Github | October 2017<br>
<br>
https://ift.tt/2xFqXkF<br>
<br>
Baishakhi Ray, Daryl Posnett, Premkumar Devanbu, Vladimir Filkov<br>
<br>
<div>
<div>
<div>
<section>
<p><span>Research highlights</span></p>
<hr>
<div>
<span>By Baishakhi Ray, Daryl Posnett, Premkumar Devanbu, Vladimir Filkov<br>
Communications of the ACM, October 2017, Vol. 60 No. 10, Pages 91-100<br>
10.1145/3126905<br>
<a href="#comments">Comments</a></span>
<div>
<figure><img alt="A Large-Scale Study of Programming Languages, illustration" title="A Large-Scale Study of Programming Languages, illustration" src="https://cacm.acm.org/system/assets/0002/8759/092117_Getty_Large-Scale-Study1.large.jpg?1506007488&amp;1506007487">
<figcaption>
<p>Credit: Getty Images</p>
</figcaption>
</figure>
</div>
<p>What is the effect of programming languages on software quality? This question has been a topic of much debate for a very long time. In this study, we gather a very large data set from GitHub (728 projects, 63 million SLOC, 29,000 authors, 1.5 million commits, in 17 languages) in an attempt to shed some empirical light on this question. This reasonably large sample size allows us to use a mixed-methods approach, combining multiple regression modeling with visualization and text analytics, to study the effect of language features such as static versus dynamic typing and allowing versus disallowing type confusion on software quality. By triangulating findings from different methods, and controlling for confounding effects such as team size, project size, and project history, we report that language design does have a significant, but modest effect on software quality. Most notably, it does appear that disallowing type confusion is modestly better than allowing it, and among functional languages, static typing is also somewhat better than dynamic typing. We also find that functional languages are somewhat better than procedural languages. It is worth noting that these modest effects arising from language design are overwhelmingly dominated by the process factors such as project size, team size, and commit size. However, we caution the reader that even these modest effects might quite possibly be due to other, intangible process factors, for example, the preference of certain personality types for functional, static languages that disallow type confusion.</p>
<p><a href="#PageTop">Back to Top</a></p>
<h3>1. Introduction</h3>
<p>A variety of debates ensue during discussions whether a given programming language is "the right tool for the job." While some of these debates may appear to be tinged with an almost religious fervor, most agree that programming language choice can impact both the coding process and the resulting artifact.</p>
<p>Advocates of strong, static typing tend to believe that the static approach catches defects early; for them, an ounce of prevention is worth a pound of cure. Dynamic typing advocates argue, however, that conservative static type checking is wasteful of developer resources, and that it is better to rely on strong dynamic type checking to catch type errors as they arise. These debates, however, have largely been of the armchair variety, supported only by anecdotal evidence.</p>
<p>This is perhaps not unreasonable; obtaining empirical evidence to support such claims is a challenging task given the number of other factors that influence software engineering outcomes, such as code quality, language properties, and usage domains. Considering, for example, software quality, there are a number of well-known influential factors, such as code size,<sup><a href="#R6">6</a></sup> team size,<sup><a href="#R2">2</a></sup> and age/maturity.<sup><a href="#R9">9</a></sup></p>
<p>Controlled experiments are one approach to examining the impact of language choice in the face of such daunting confounds, however, owing to cost, such studies typically introduce a confound of their own, that is, limited scope. The tasks completed in such studies are necessarily limited and do not emulate <em>real world</em> development. There have been several such studies recently that use students, or compare languages with static or dynamic typing through an experimental factor.<sup><a href="#R7">7</a>, <a href="#R12">12</a>, <a href="#R15">15</a></sup></p>
<p>Fortunately, we can now study these questions over a large body of real-world software projects. GitHub contains many projects in multiple languages that substantially vary across size, age, and number of developers. Each project repository provides a detailed record, including contribution history, project size, authorship, and defect repair. We then use a variety of tools to study the effects of language features on defect occurrence. Our approach is best described as mixed-methods, or triangulation<sup><a href="#R5">5</a></sup> approach; we use text analysis, clustering, and visualization to confirm and support the findings of a quantitative regression study. This empirical approach helps us to understand the practical impact of programming languages, as they are used colloquially by developers, on software quality.</p>
<p><a href="#PageTop">Back to Top</a></p>
<h3>2. Methodology</h3>
<p>Our methods are typical of large scale observational studies in software engineering. We first gather our data from several sources using largely automated methods. We then filter and clean the data in preparation for building a statistical model. We further validate the model using qualitative methods. Filtering choices are driven by a combination of factors including the nature of our research questions, the quality of the data and beliefs about which data is most suitable for statistical study. In particular, GitHub contains many projects written in a large number of programming languages. For this study, we focused our data collection efforts on the most popular projects written in the most popular languages. We choose statistical methods appropriate for evaluating the impact of factors on count data.</p>
<p><img alt="*" src="https://dl.acm.org/images/bullet.gif">&nbsp;<strong>2.1. Data collection</strong></p>
<p>We choose the top 19 programming languages from GitHub. We disregard CSS, Shell script, and Vim script as they are not considered to be general purpose languages. We further include <code>Typescript</code>, a typed superset of <code>JavaScript</code>. Then, for each of the studied languages we retrieve the top 50 projects that are primarily written in that language. In total, we analyze 850 projects spanning 17 different languages.</p>
<p>Our language and project data was extracted from the <em>GitHub Archive</em>, a database that records all public GitHub activities. The archive logs 18 different GitHub events including new commits, fork events, pull request, developers' information, and issue tracking of all the open source GitHub projects on an hourly basis. The archive data is uploaded to Google BigQuery to provide an interface for interactive data analysis.</p>
<p><strong>Identifying top languages.</strong> We aggregate projects based on their primary language. Then we select the languages with the most projects for further analysis, as shown in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t1.jpg">Table 1</a>. A given project can use many languages; assigning a single language to it is difficult. Github Archive stores information gathered from GitHub Linguist which measures the language distribution of a project repository using the source file extensions. The language with the maximum number of source files is assigned as the <em>primary language</em> of the project.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t1.jpg"><img alt="t1.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/t1.jpg" height="350"></a><br>
<strong>Table 1. Top 3 projects in each language.</strong></p>
<p><strong>Retrieving popular projects.</strong> For each selected language, we filter the project repositories written primarily in that language by its popularity based on the associated number of <em>stars.</em> This number indicates how many people have actively expressed interest in the project, and is a reasonable proxy for its popularity. Thus, the top 3 projects in C are <em>linux, git</em>, and <em>php-src</em>; and for C++ they are <em>node-webkit, phantomjs</em>, and <em>mongo</em>; and for <code>Java</code> they are <em>storm, elasticsearch</em>, and <em>ActionBarSherlock.</em> In total, we select the top 50 projects in each language.</p>
<p>To ensure that these projects have a sufficient development history, we drop the projects with fewer than 28 commits (28 is the first quartile commit count of considered projects). This leaves us with 728 projects. <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t1.jpg">Table 1</a> shows the top 3 projects in each language.</p>
<p><strong>Retrieving project evolution history.</strong> For each of 728 projects, we downloaded the non-merged commits, commit logs, author date, and author name using <em>git.</em> We compute code churn and the number of files modified per commit from the number of added and deleted lines per file. We retrieve the languages associated with each commit from the extensions of the modified files (a commit can have multiple language tags). For each commit, we calculate its <em>commit age</em> by subtracting its commit date from the first commit of the corresponding project. We also calculate other project-related statistics, including maximum commit age of a project and the total number of developers, used as control variables in our regression model, and discussed in Section 3. We identify bug fix commits made to individual projects by searching for error related keywords: "error," "bug," "fix," "issue," "mistake," "incorrect," "fault," "defect," and "flaw," in the commit log, similar to a prior study.<sup><a href="#R18">18</a></sup></p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t2.jpg">Table 2</a> summarizes our data set. Since a project may use multiple languages, the second column of the table shows the total number of projects that use a certain language at some capacity. We further exclude some languages from a project that have fewer than 20 commits in that language, where 20 is the first quartile value of the total number of commits per project per language. For example, we find 220 projects that use more than 20 commits in C. This ensures sufficient activity for each language&ndash;project pair.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t2.jpg"><img alt="t2.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/t2.jpg" height="364"></a><br>
<strong>Table 2. Study subjects.</strong></p>
<p>In summary, we study 728 projects developed in 17 languages with 18 years of history. This includes 29,000 different developers, 1.57 million commits, and 564,625 bug fix commits.</p>
<p><img alt="*" src="https://dl.acm.org/images/bullet.gif">&nbsp;<strong>2.2. Categorizing languages</strong></p>
<p>We define language classes based on several properties of the language thought to influence language quality,<sup><a href="#R7">7</a>, <a href="#R8">8</a>, <a href="#R12">12</a></sup> as shown in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t3.jpg">Table 3</a>. The <em>Programming Paradigm</em> indicates whether the project is written in an imperative procedural, imperative scripting, or functional language. In the rest of the paper, we use the terms procedural and scripting to indicate imperative procedural and imperative scripting respectively.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t3.jpg"><img alt="t3.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/t3.jpg" height="389"></a><br>
<strong>Table 3. Different types of language classes.</strong></p>
<p><em>Type Checking</em> indicates static or dynamic typing. In statically typed languages, type checking occurs at compile time, and variable names are bound to a value and to a type. In addition, expressions (including variables) are classified by types that correspond to the values they might take on at run-time. In dynamically typed languages, type checking occurs at run-time. Hence, in the latter, it is possible to bind a variable name to objects of different types in the same program.</p>
<p><em>Implicit Type Conversion</em> allows access of an operand of type T1 as a different type T2, without an explicit conversion. Such implicit conversion may introduce type-confusion in some cases, especially when it presents an operand of specific type T1, as an instance of a different type T2. Since not all implicit type conversions are immediately a problem, we operationalize our definition by showing examples of the implicit type confusion that can happen in all the languages we identified as allowing it. For example, in languages like <code>Perl, JavaScript</code>, and <code>CoffeeScript</code> adding a string to a number is permissible (e.g., "5" + 2 yields "52"). The same operation yields 7 in <code>Php</code>. Such an operation is not permitted in languages such as <code>Java</code> and <code>Python</code> as they do not allow implicit conversion. In C and C++ coercion of data types can result in unintended results, for example, <code>int x; float y; y=3.5; x=y</code>; is legal C code, and results in different values for x and y, which, depending on intent, may be a problem downstream.<sup><a href="#FNA">a</a></sup> In <code>Objective-C</code> the data type <em>id</em> is a generic object pointer, which can be used with an object of any data type, regardless of the class.<sup><a href="#FNB">b</a></sup> The flexibility that such a generic data type provides can lead to implicit type conversion and also have unintended consequences.<sup><a href="#FNC">c</a></sup> Hence, we classify a language based on whether its compiler <em>allows</em> or <em>disallows</em> the implicit type conversion as above; the latter explicitly detects type confusion and reports it.</p>
<p>Disallowing implicit type conversion could result from static type inference within a compiler (e.g., with <code>Java</code>), using a type-inference algorithm such as Hindley<sup><a href="#R10">10</a></sup> and Milner,<sup><a href="#R17">17</a></sup> or at run-time using a dynamic type checker. In contrast, a type-confusion can occur silently because it is either undetected or is unreported. Either way, implicitly allowing type conversion provides flexibility but may eventually cause errors that are difficult to localize. To abbreviate, we refer to languages allowing implicit type conversion as <em>implicit</em> and those that disallow it as <em>explicit.</em></p>
<p><em>Memory Class</em> indicates whether the language requires developers to manage memory. We treat <code>Objective-C</code> as unmanaged, in spite of it following a hybrid model, because we observe many memory errors in its codebase, as discussed in RQ4 in Section 3.</p>
<p>Note that we classify and study the languages as they are colloquially used by developers in real-world software. For example, <code>TypeScript</code> is intended to be used as a static language, which disallows implicit type conversion. However, in practice, we notice that developers often (for 50% of the variables, and across <code>TypeScript</code>-using projects in our dataset) use the <code>any</code> type, a catch-all union type, and thus, in practice, <code>TypeScript</code> allows dynamic, implicit type conversion. To minimize the confusion, we exclude <code>TypeScript</code> from our language classifications and the corresponding model (see <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t3.jpg">Table 3</a> and <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t7.jpg">7</a>).</p>
<p><img alt="*" src="https://dl.acm.org/images/bullet.gif">&nbsp;<strong>2.3. Identifying project domain</strong></p>
<p>We classify the studied projects into different domains based on their features and function using a mix of automated and manual techniques. The projects in GitHub come with <code>project descriptions</code> and README files that describe their features. We used Latent Dirichlet Allocation (LDA)<sup><a href="#R3">3</a></sup> to analyze this text. Given a set of documents, LDA identifies a set of topics where each topic is represented as probability of generating different words. For each document, LDA also estimates the probability of assigning that document to each topic.</p>
<p>We detect 30 distinct domains, that is, topics, and estimate the probability that each project belonging to each domain. Since these auto-detected domains include several project-specific keywords, for example, facebook, it is difficult to identify the underlying common functions. In order to assign a meaningful name to each domain, we manually inspect each of the 30 domains to identify projectname-independent, domain-identifying keywords. We manually rename all of the 30 auto-detected domains and find that the majority of the projects fall under six domains: Application, Database, CodeAnalyzer, Middleware, Library, and Framework. We also find that some projects do not fall under any of the above domains and so we assign them to a catchall domain labeled as <em>Other</em>. This classification of projects into domains was subsequently checked and confirmed by another member of our research group. <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t4.jpg">Table 4</a> summarizes the identified domains resulting from this process.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t4.jpg"><img alt="t4.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/t4.jpg" height="206"></a><br>
<strong>Table 4. Characteristics of domains.</strong></p>
<p><img alt="*" src="https://dl.acm.org/images/bullet.gif">&nbsp;<strong>2.4. Categorizing bugs</strong></p>
<p>While fixing software bugs, developers often leave important information in the commit logs about the nature of the bugs; for example, why the bugs arise and how to fix the bugs. We exploit such information to categorize the bugs, similar to Tan <em>et al.</em><sup><a href="#R13">13</a>, <a href="#R24">24</a></sup></p>
<p>First, we categorize the bugs based on their <em>Cause</em> and <em>Impact. Causes</em> are further classified into disjoint subcategories of errors: Algorithmic, Concurrency, Memory, generic Programming, and Unknown. The bug <em>Impact</em> is also classified into four disjoint subcategories: Security, Performance, Failure, and Other unknown categories. Thus, each bug-fix commit also has an induced Cause and an Impact type. <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t5.jpg">Table 5</a> shows the description of each bug category. This classification is performed in two phases:</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t5.jpg"><img alt="t5.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/t5.jpg" height="117"></a><br>
<strong>Table 5. Categories of bugs and their distribution in the whole dataset.</strong></p>
<p><strong>(1) Keyword search.</strong> We randomly choose 10% of the bug-fix messages and use a keyword based search technique to automatically categorize them as potential bug types. We use this annotation, separately, for both Cause and Impact types. We chose a restrictive set of keywords and phrases, as shown in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t5.jpg">Table 5</a>. Such a restrictive set of keywords and phrases helps reduce false positives.</p>
<p><strong>(2) Supervised classification.</strong> We use the annotated bug fix logs from the previous step as training data for supervised learning techniques to classify the remainder of the bug fix messages by treating them as test data. We first convert each bug fix message to a bag-of- words. We then remove words that appear only once among all of the bug fix messages. This reduces project specific keywords. We also stem the bag-of- words using standard natural language processing techniques. Finally, we use Support Vector Machine to classify the test data.</p>
<p>To evaluate the accuracy of the bug classifier, we manually annotated 180 randomly chosen bug fixes, equally distributed across all of the categories. We then compare the result of the automatic classifier with the manually annotated data set. The performance of this process was acceptable with precision ranging from a low of 70% for performance bugs to a high of 100% for concurrency bugs with an average of 84%. Recall ranged from 69% to 91% with an average of 84%.</p>
<p>The result of our bug classification is shown in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t5.jpg">Table 5</a>. Most of the defect causes are related to generic programming errors. This is not surprising as this category involves a wide variety of programming errors such as type errors, typos, compilation error, etc. Our technique could not classify 1.04% of the bug fix messages in any Cause or Impact category; we classify these as Unknown.</p>
<p><img alt="*" src="https://dl.acm.org/images/bullet.gif">&nbsp;<strong>2.5. Statistical methods</strong></p>
<p>We model the number of defective commits against other factors related to software projects using regression. All models use <em>negative binomial regression</em> (NBR) to model the counts of project attributes such as the number of commits. NBR is a type of generalized linear model used to model non-negative integer responses.<sup><a href="#R4">4</a></sup></p>
<p>In our models we control for several language per-project dependent factors that are likely to influence the outcome. Consequently, each (language, project) pair is a row in our regression and is viewed as a sample from the population of open source projects. We log-transform dependent count variables as it stabilizes the variance and usually improves the model fit.<sup><a href="#R4">4</a></sup> We verify this by comparing transformed with non transformed data using the AIC and Vuong's test for non-nested models.</p>
<p>To check that excessive multicollinearity is not an issue, we compute the variance inflation factor of each dependent variable in all of the models with a conservative maximum value of 5.<sup><a href="#R4">4</a></sup> We check for and remove high leverage points through visual examination of the residuals versus leverage plot for each model, looking for both separation and large values of Cook's distance.</p>
<p>We employ <em>effects</em>, or <em>contrast</em>, coding in our study to facilitate interpretation of the language coefficients.<sup><a href="#R4">4</a></sup> Weighted effects codes allow us to compare each language to the average effect across all languages while compensating for the unevenness of language usage across projects.<sup><a href="#R23">23</a></sup> To test for the relationship between two factor variables we use a Chi-square test of independence.<sup><a href="#R14">14</a></sup> After confirming a dependence we use Cramer's V, an <em>r</em> &times; <em>c</em> equivalent of the phi coefficient for nominal data, to establish an effect size.</p>
<p><a href="#PageTop">Back to Top</a></p>
<h3>3. Results</h3>
<p>We begin with a straightforward question that directly addresses the core of what some fervently believe must be true, namely:</p>
<p><strong>RQ1. Are some languages more defect-prone than others?</strong></p>
<p>We use a regression model to compare the impact of each language on the number of defects with the average impact of all languages, against defect fixing commits (see <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t6.jpg">Table 6</a>).</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t6.jpg"><img alt="t6.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/t6.jpg" height="439"></a><br>
<strong>Table 6. Some languages induce fewer defects than other languages.</strong></p>
<p>We include some variables as controls for factors that will clearly influence the response. Project age is included as older projects will generally have a greater number of defect fixes. Trivially, the number of commits to a project will also impact the response. Additionally, the number of developers who touch a project and the raw size of the project are both expected to grow with project activity.</p>
<p>The sign and magnitude of the estimated coefficients in the above model relates the predictors to the outcome. The first four variables are control variables and we are not interested in their impact on the outcome other than to say that they are all positive and significant. The language variables are indicator variables, viz. factor variables, for each project. The coefficient compares each language to the grand weighted mean of all languages in all projects. The language coefficients can be broadly grouped into three general categories. The first category is those for which the coefficient is statistically insignificant and the modeling procedure could not distinguish the coefficient from zero. These languages may behave similar to the average or they may have wide variance. The remaining coefficients are significant and either positive or negative. For those with positive coefficients we can expect that the language is associated with a greater number of defect fixes. These languages include <code>C, C++, Objective-C, Php</code>, and <code>Python</code>. The languages <code>Clojure, Haskell, Ruby</code>, and <code>Scala</code>, all have negative coefficients implying that these languages are less likely than average to result in defect fixing commits.</p>
<p>One should take care not to overestimate the impact of language on defects. While the observed relationships are statistically significant, the effects are quite small. Analysis of deviance reveals that language accounts for less than 1% of the total explained deviance.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/ut1.jpg"><img alt="ut1.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/ut1.jpg" height="127"></a></p>
<p>We can read the model coefficients as the expected change in the log of the response for a one unit change in the predictor with all other predictors held constant; that is, for a coefficient <em>&beta;<sub>i</sub></em>, a one unit change in <em>&beta;<sub>i</sub></em> yields an expected change in the response of e<sup><em>&beta;i</em></sup>. For the factor variables, this expected change is compared to the average across all languages. Thus, if, for some number of commits, a particular project developed in an <em>average</em> language had four defective commits, then the choice to use C++ would mean that we should expect one additional defective commit since e<sup>0.18</sup> &times; 4 = 4.79. For the same project, choosing <code>Haskell</code> would mean that we should expect about one fewer defective commit as <em>e</em><sup>&minus;0.26</sup> &times; 4 = 3.08. The accuracy of this prediction depends on all other factors remaining the same, a challenging proposition for all but the most trivial of projects. All observational studies face similar limitations; we address this concern in more detail in Section 5.</p>
<p><strong>Result 1:</strong> <em>Some languages have a greater association with defects than other languages, although the effect is small.</em></p>
<p>In the remainder of this paper we expand on this basic result by considering how different categories of application, defect, and language, lead to further insight into the relationship between languages and defect proneness.</p>
<p>Software bugs usually fall under two broad categories: (1) <em>Domain Specific bug</em>: specific to project function and do not depend on the underlying programming language. (2) <em>Generic bug</em>: more generic in nature and has less to do with project function, for example, typeerrors, concurrency errors, etc.</p>
<p>Consequently, it is reasonable to think that the interaction of application domain and language might impact the number of defects within a project. Since some languages are believed to excel at some tasks more so than others, for example, C for low level work, or <code>Java</code> for user applications, making an inappropriate choice might lead to a greater number of defects. To study this we should ideally ignore the domain specific bugs, as generic bugs are more likely to depend on the programming language featured. However, since a domain-specific bugs may also arise due to a generic programming error, it is difficult to separate the two. A possible workaround is to study languages while controlling the domain. Statistically, however, with 17 languages across 7 domains, the large number of terms would be challenging to interpret given the sample size.</p>
<p>Given this, we first consider testing for the dependence between domain and language usage within a project, using a Chi-square test of independence. Of 119 cells, 46, that is, 39%, are below the value of 5 which is too high. No more than 20% of the counts should be below 5.<sup><a href="#R14">14</a></sup> We include the value here for completeness<sup><a href="#FND">d</a></sup>; however, the low strength of association of 0.191 as measured by Cramer's V, suggests that any relationship between domain and language is small and that inclusion of domain in regression models would not produce meaningful results.</p>
<p>One option to address this concern would be to remove languages or combine domains, however, our data here presents no clear choices. Alternatively, we could combine languages; this choice leads to a related but slightly different question.</p>
<p><strong>RQ2. Which language properties relate to defects?</strong></p>
<p>Rather than considering languages individually, we aggregate them by language class, as described in Section 2.2, and analyze the relationship to defects. Broadly, each of these properties divides languages along lines that are often discussed in the context of errors, drives user debate, or has been the subject of prior work. Since the individual properties are highly correlated, we create six model factors that combine all of the individual factors across all of the languages in our study. We then model the impact of the six different factors on the number of defects while controlling for the same basic covariates that we used in the model in <em>RQ1</em>.</p>
<p>As with language (earlier in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t6.jpg">Table 6</a>), we are comparing language <em>classes</em> with the average behavior across all language classes. The model is presented in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t7.jpg">Table 7</a>. It is clear that <code>Script-Dynamic-Explicit-Managed</code> class has the smallest magnitude coefficient. The coefficient is insignificant, that is, the z-test for the coefficient cannot distinguish the coefficient from zero. Given the magnitude of the standard error, however, we can assume that the behavior of languages in this class is very close to the average across all languages. We confirm this by recoding the coefficient using <code>Proc-Static-Implicit-Unmanaged</code> as the base level and employing treatment, or dummy coding that compares each language class with the base level. In this case, <code>Script-Dynamic-Explicit-Managed</code> is significantly different with <em>p</em> = 0.00044. We note here that while choosing different coding methods affects the coefficients and z-scores, the models are identical in all other respects. When we change the coding we are rescaling the coefficients to reflect the comparison that we wish to make.<sup><a href="#R4">4</a></sup> Comparing the other language classes to the grand mean, <code>Proc-Static-Implicit-Unmanaged</code> languages are more likely to induce defects. This implies that either implicit type conversion or memory management issues contribute to greater defect proneness as compared with other procedural languages.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t7.jpg"><img alt="t7.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/t7.jpg" height="260"></a><br>
<strong>Table 7. Functional languages have a smaller relationship to defects than other language classes whereas procedural languages are greater than or similar to the average.</strong></p>
<p>Among scripting languages we observe a similar relationship between languages that allow versus those that do not allow implicit type conversion, providing some evidence that implicit type conversion (vs. explicit) is responsible for this difference as opposed to memory management. We cannot state this conclusively given the correlation between factors. However when compared to the average, as a group, languages that do not allow implicit type conversion are less error-prone while those that do are more error-prone. The contrast between static and dynamic typing is also visible in functional languages.</p>
<p>The functional languages as a group show a strong difference from the average. Statically typed languages have a substantially smaller coefficient yet both functional language classes have the same standard error. This is strong evidence that functional static languages are less error-prone than functional dynamic languages, however, the z-tests only test whether the coefficients are different from zero. In order to strengthen this assertion, we recode the model as above using treatment coding and observe that the <code>Functional-Static-Explicit-Managed</code> language class is significantly less defect-prone than the <code>Functional-Dynamic-Explicit-Managed</code> language class with <em>p</em> = 0.034.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/ut2.jpg"><img alt="ut2.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/ut2.jpg" height="127"></a></p>
<p>As with language and defects, the relationship between language class and defects is based on a small effect. The deviance explained is similar, albeit smaller, with language class explaining much less than 1% of the deviance.</p>
<p>We now revisit the question of application domain. Does domain have an interaction with language class? Does the choice of, for example, a functional language, have an advantage for a particular domain? As above, a Chi-square test for the relationship between these factors and the project domain yields a value of 99.05 and <em>df</em> = 30 with <em>p</em> = 2.622e&ndash;09 allowing us to reject the null hypothesis that the factors are independent. Cramer's-V yields a value of 0.133, a weak level of association. Consequently, although there is some relation between domain and language, there is only a weak relationship between domain and language class.</p>
<p><strong>Result 2:</strong> <em>There is a small but significant relationship between language class and defects. Functional languages are associated with fewer defects than either procedural or scripting languages.</em></p>
<p>It is somewhat unsatisfying that we do not observe a strong association between language, or language class, and domain within a project. An alternative way to view this same data is to disregard projects and aggregate defects over all languages and domains. Since this does not yield independent samples, we do not attempt to analyze it statistically, rather we take a descriptive, visualization-based approach.</p>
<p>We define <em>Defect Proneness</em> as the ratio of bug fix commits over total commits per language per domain. <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/f1.jpg">Figure 1</a> illustrates the interaction between domain and language using a heat map, where the defect proneness increases from lighter to darker zone. We investigate which language factors influence defect fixing commits across a collection of projects written across a variety of languages. This leads to the following research question:</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/f1.jpg"><img alt="f1.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/f1.jpg" height="285"></a><br>
<strong>Figure 1. Interaction of language's defect proneness with domain. Each cell in the heat map represents defect proneness of a language (row header) for a given domain (column header). The "Overall" column represents defect proneness of a language over all the domains. The cells with white cross mark indicate null value, that is, no commits were made corresponding to that cell.</strong></p>
<p><strong>RQ3. Does language defect proneness depend on domain?</strong></p>
<p>In order to answer this question we first filtered out projects that would have been viewed as outliers, filtered as high leverage points, in our regression models. This was necessary here as, even though this is a nonstatistical method, some relationships could impact visualization. For example, we found that a single project, Google's v8, a <code>JavaScript</code> project, was responsible for all of the errors in Middleware. This was surprising to us since <code>JavaScript</code> is typically not used for Middleware. This pattern repeats in other domains, consequently, we filter out the projects that have defect density below 10 and above 90 percentile. The result is in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/f1.jpg">Figure 1</a>.</p>
<p>We see only a subdued variation in this heat map which is a result of the inherent defect proneness of the languages as seen in RQ1. To validate this, we measure the pairwise rank correlation between the language defect proneness for each domain with the overall. For all of the domains except Database, the correlation is positive, and p-values are significant (&lt;0.01). Thus, w.r.t. defect proneness, the language ordering in each domain is strongly correlated with the overall language ordering.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/ut3.jpg"><img alt="ut3.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/ut3.jpg" height="63"></a></p>
<p><strong>Result 3:</strong> <em>There is no general relationship between application domain and language defect proneness.</em></p>
<p>We have shown that different languages induce a larger number of defects and that this relationship is not only related to particular languages but holds for general classes of languages; however, we find that the type of project does not mediate this relationship to a large degree. We now turn our attention to categorization of the response. We want to understand how language relates to specific kinds of defects and how this relationship compares to the more general relationship that we observe. We divide the defects into categories as described in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t5.jpg">Table 5</a> and ask the following question:</p>
<p><strong>RQ4. What is the relation between language and bug category?</strong></p>
<p>We use an approach similar to RQ3 to understand the relation between languages and bug categories. First, we study the relation between bug categories and language class. A heat map (<a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/f2.jpg">Figure 2</a>) shows aggregated defects over language classes and bug types. To understand the interaction between bug categories and languages, we use an NBR regression model for each category. For each model we use the same control factors as RQ1 as well as languages encoded with weighted effects to predict defect fixing commits.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/f2.jpg"><img alt="f2.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/f2.jpg" height="233"></a><br>
<strong>Figure 2. Relation between bug categories and language class. Each cell represents percentage of bug fix commit out of all bug fix commits per language class (row header) per bug category (column header). The values are normalized column wise.</strong></p>
<p>The results along with the anova value for language are shown in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t8.jpg">Table 8</a>. The overall deviance for each model is substantially smaller and the proportion explained by language for a specific defect type is similar in magnitude for most of the categories. We interpret this relationship to mean that language has a greater impact on specific categories of bugs, than it does on bugs overall. In the next section we expand on these results for the bug categories with significant bug counts as reported in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t5.jpg">Table 5</a>. However, our conclusion generalizes for all categories.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t8.jpg"><img alt="t8.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/t8.jpg" height="250"></a><br>
<strong>Table 8. While the impact of language on defects varies across defect category, language has a greater impact on specific categories than it does on defects in general.</strong></p>
<p><strong>Programming errors.</strong> Generic programming errors account for around 88.53% of all bug fix commits and occur in all the language classes. Consequently, the regression analysis draws a similar conclusion as of RQ1 (see <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t6.jpg">Table 6</a>). All languages incur programming errors such as faulty error-handling, faulty definitions, typos, etc.</p>
<p><strong>Memory errors.</strong> Memory errors account for 5.44% of all the bug fix commits. The heat map in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/f2.jpg">Figure 2</a> shows a strong relationship between <code>Proc-Static-Implicit-Unmanaged</code> class and memory errors. This is expected as languages with unmanaged memory are known for memory bugs. <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t8.jpg">Table 8</a> confirms that such languages, for example, C, C++, and <code>Objective-C</code> introduce more memory errors. Among the managed languages, <code>Java</code> induces more memory errors, although fewer than the unmanaged languages. Although <code>Java</code> has its own garbage collector, memory leaks are not surprising since unused object references often prevent the garbage collector from reclaiming memory.<sup><a href="#R11">11</a></sup> In our data, 28.89% of all the memory errors in <code>Java</code> are the result of a memory leak. In terms of effect size, language has a larger impact on memory defects than all other <em>cause</em> categories.</p>
<p><strong>Concurrency errors.</strong> 1.99% of the total bug fix commits are related to concurrency errors. The heat map shows that <code>Proc-Static-Implicit-Unmanaged</code> dominates this error type. C and C++ introduce 19.15% and 7.89% of the errors, and they are distributed across the projects.</p>
<p><a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/ut4.jpg"><img alt="ut4.jpg" width="415" src="http://deliveryimages.acm.org/10.1145/3130000/3126905/ut4.jpg" height="95"></a></p>
<p>Both of the <code>Static-Strong-Managed</code> language classes are in the darker zone in the heat map confirming, in general static languages produce more concurrency errors than others. Among the dynamic languages, only <code>Erlang</code> is more prone to concurrency errors, perhaps relating to the greater use of this language for concurrent applications. Likewise, the negative coefficients in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t8.jpg">Table 8</a> shows that projects written in dynamic languages like <code>Ruby</code> and <code>Php</code> have fewer concurrency errors. Note that, certain languages like <code>JavaScript, CoffeeScript</code>, and <code>TypeScript</code> do not support concurrency, in its traditional form, while <code>Php</code> has a limited support depending on its implementations. These languages introduce artificial zeros in the data, and thus the concurrency model coefficients in <a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t8.jpg">Table 8</a> for those languages cannot be interpreted like the other coefficients. Due to these artificial zeros, the average over all languages in this model is smaller, which may affect the sizes of the coefficients, since they are given w.r.t. the average, but it will not affect their relative relationships, which is what we are after.</p>
<p>A textual analysis based on word-frequency of the bug fix messages suggests that most of the concurrency errors occur due to a race condition, deadlock, or incorrect synchronization, as shown in the table above. Across all language, race conditions are the most frequent cause of such errors, for example, 92% in <code>Go</code>. The enrichment of race condition errors in <code>Go</code> is probably due to an accompanying race-detection tool that may help developers locate races. The synchronization errors are primarily related to message passing interface (MPI) or shared memory operation (SHM). <code>Erlang</code> and <code>Go</code> use MPI<sup><a href="#FNE">e</a></sup> for inter-thread communication, which explains why these two languages do not have any SHM related errors such as locking, mutex, etc. In contrast, projects in the other languages use SHM primitives for communication and can thus may have locking-related errors.</p>
<p><strong>Security and other impact errors.</strong> Around 7.33% of all the bug fix commits are related to Impact errors. Among them <code>Erlang, C++</code>, and <code>Python</code> associate with more security errors than average (<a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/t8.jpg">Table 8</a>). <code>Clojure</code> projects associate with fewer security errors (<a href="http://deliveryimages.acm.org/10.1145/3130000/3126905/f2.jpg">Figure 2</a>). From the heat map we also see that <code>Static</code> languages are in general more prone to failure and performance errors, these are followed by <code>Functional-Dynamic-Explicit-Managed</code> languages such as <code>Erlang</code>. The analysis of deviance results confirm that language is strongly associated with failure impacts. While security errors are the weakest among the categories, the deviance explained by language is still quite strong when compared with the residual deviance.</p>
<p><strong>Result 4:</strong> <em>Defect types are strongly associated with languages; some defect type like memory errors and concurrency errors also depend on language primitives. Language matters more for specific categories than it does for defects overall.</em></p>
<p><a href="#PageTop">Back to Top</a></p>
<h3>4. Related Work</h3>
<p>Prior work on programming language comparison falls in three categories:</p>
<p><strong>(1) <em>Controlled experiment.</em></strong> For a given task, developers are monitored while programming in different languages. Researchers then compare outcomes such as development effort and program quality. Hanenberg<sup><a href="#R7">7</a></sup> compared static versus dynamic typing by monitoring 48 programmers for 27 h while developing a parser program. He found no significant difference in code quality between the two; however, dynamic type-based languages were found to have shorter development time. Their study was conducted with undergraduate students in a lab setting with custom-designed language and IDE. Our study, by contrast is a field study of popular software applications. While we can only indirectly (and <em>post facto</em>) control for confounding factors using regression, we benefit from much larger sample sizes, and more realistic, widely-used software. We find that statically typed languages in general are less defect-prone than the dynamic types, and that disallowing implicit type conversion is better than allowing it, in the same regard. The effect sizes are modest; it could be reasonably argued that they are visible here precisely because of the large sample sizes.</p>
<p>Harrison et al.<sup><a href="#R8">8</a></sup> compared C++, a procedural language, with <code>SML</code>, a functional language, finding no significant difference in total number of errors, although <code>SML</code> has higher defect density than C++. <code>SML</code> was not represented in our data, which however, suggest that functional languages are generally less defect-prone than procedural languages. Another line of work primarily focuses on comparing development effort across different languages.<sup><a href="#R12">12</a>, <a href="#R20">20</a></sup> However, they do not analyze language defect proneness.</p>
<p><strong>(2) <em>Surveys.</em></strong> Meyerovich and Rabkin<sup><a href="#R16">16</a></sup> survey developers' views of programming languages, to study why some languages are more popular than others. They report strong influence from non-linguistic factors: prior language skills, availability of open source tools, and existing legacy systems. Our study also confirms that the availability of external tools also impacts software quality; for example, concurrency bugs in <code>Go</code> (see RQ4 in Section 3).</p>
<p><strong>(3) <em>Repository mining.</em></strong> Bhattacharya and Neamtiu<sup><a href="#R1">1</a></sup> study four projects developed in both C and C++ and find that the software components developed in C++ are in general more reliable than C. We find that both C and C++ are more defect-prone than average. However, for certain bug types like concurrency errors, C is more defect-prone than C++ (see RQ4 in Section 3).</p>
<p><a href="#PageTop">Back to Top</a></p>
<h3>5. Threats to Validity</h3>
<p>We recognize few threats to our reported results. First, to identify bug fix commits we rely on the keywords that developers often use to indicate a bug fix. Our choice was deliberate. We wanted to capture the issues that developers continuously face in an ongoing development process, rather than reported bugs. However, this choice possesses threats of over estimation. Our categorization of domains is subject to interpreter bias, although another member of our group verified the categories. Also, our effort to categorize bug fix commits could potentially be tainted by the initial choice of keywords. The descriptiveness of commit logs vary across projects. To mitigate these threats, we evaluate our classification against manual annotation as discussed in Section 2.4.</p>
<p>We determine the language of a file based on its extension. This can be error-prone if a file written in a different language takes a common language extension that we have studied. To reduce such error, we manually verified language categorization against a randomly sampled file set.</p>
<p>To interpret language class in Section 2.2, we make certain assumptions based on how a language property is most commonly used, as reflected in our data set, for example, we classify <code>Objective-C</code> as unmanaged memory type rather than hybrid. Similarly, we annotate <code>Scala</code> as functional and C# as procedural, although both support either design choice.<sup><a href="#R19">19</a>, <a href="#R21">21</a></sup> We do not distinguish object-oriented languages (OOP) from procedural languages in this work as there is no clear distinction, the difference largely depends on programming style. We categorize C++ as allowing implicit type conversion because a memory region of a certain type can be treated differently using pointer manipulation.<sup><a href="#R22">22</a></sup> We note that most C++ compilers can detect type errors at compile time.</p>
<p>Finally, we associate defect fixing commits to language properties, although they could reflect reporting style or other developer properties. Availability of external tools or libraries may also impact the extent of bugs associated with a language.</p>
<p><a href="#PageTop">Back to Top</a></p>
<h3>6. Conclusion</h3>
<p>We have presented a large-scale study of language type and use as it relates to software quality. The Github data we used is characterized by its complexity and variance along multiple dimensions. Our sample size allows a mixed-methods study of the effects of language, and of the interactions of language, domain, and defect type while controlling for a number of confounds. The data indicates that functional languages are better than procedural languages; it suggests that disallowing implicit type conversion is better than allowing it; that static typing is better than dynamic; and that managed memory usage is better than unmanaged. Further, that the defect proneness of languages in general is not associated with software domains. Additionally, languages are more related to individual bug categories than bugs overall.</p>
<p>On the other hand, even large datasets become small and insufficient when they are sliced and diced many ways simultaneously. Consequently, with an increasing number of dependent variables it is difficult to answer questions about a specific variable's effect, especially where variable interactions exist. Hence, we are unable to quantify the specific effects of language type on usage. Additional methods such as surveys could be helpful here. Addressing these challenges remains for future work.</p>
<p><a href="#PageTop">Back to Top</a></p>
<h3>Acknowledgments</h3>
<p>This material is based upon work supported by the National Science Foundation under grant nos. 1445079, 1247280, 1414172, 1446683 and from AFOSR award FA955-11-1-0246.</p>
<p><a href="#PageTop">Back to Top</a></p>
<div>
<a name="references" id="references"></a>
<h3>References</h3>
<p><a name="R1" id="R1"></a>1. Bhattacharya, P., Neamtiu, I. Assessing programming language impact on development and maintenance: A study on C and C++. In <em>Proceedings of the 33rd International Conference on Software Engineering, ICSE'11</em> (New York, NY USA, 2011). ACM, 171&ndash;180.</p>
<p><a name="R2" id="R2"></a>2. Bird, C., Nagappan, N., Murphy, B., Gall, H., Devanbu, P. Don't touch my code! Examining the effects of ownership on software quality. In <em>Proceedings of the 19th ACM SIGSOFT Symposium and the 13th European Conference on Foundations of Software Engineering</em> (2011). ACM, 4&ndash;14.</p>
<p><a name="R3" id="R3"></a>3. Blei, D.M. Probabilistic topic models. <em>Commun. ACM 55</em>, 4 (2012), 77&ndash;84.</p>
<p><a name="R4" id="R4"></a>4. Cohen, J. <em>Applied Multiple Regression/Correlation Analysis for the Behavioral Sciences.</em> Lawrence Erlbaum, 2003.</p>
<p><a name="R5" id="R5"></a>5. Easterbrook, S., Singer, J., Storey, M.-A., Damian, D. Selecting empirical methods for software engineering research. In <em>Guide to Advanced Empirical Software Engineering</em> (2008). Springer, 285&ndash;311.</p>
<p><a name="R6" id="R6"></a>6. El Emam, K., Benlarbi, S., Goel, N., Rai, S.N. The confounding effect of class size on the validity of object-oriented metrics. <em>IEEE Trans. Softw. Eng. 27</em>, 7 (2001), 630&ndash;650.</p>
<p><a name="R7" id="R7"></a>7. Hanenberg, S. An experiment about static and dynamic type systems: Doubts about the positive impact of static type systems on development time. In <em>Proceedings of the ACM International Conference on Object Oriented Programming Systems Languages and Applications, OOPSLA'10</em> (New York, NY, USA, 2010). ACM, 22&ndash;35.</p>
<p><a name="R8" id="R8"></a>8. Harrison, R., Smaraweera, L., Dobie, M., Lewis, P. Comparing programming paradigms: An evaluation of functional and object-oriented programs. <em>Softw. Eng. J. 11</em>, 4 (1996), 247&ndash;254.</p>
<p><a name="R9" id="R9"></a>9. Harter, D.E., Krishnan, M.S., Slaughter, S.A. Effects of process maturity on quality, cycle time, and effort in software product development. <em>Manage. Sci. 46</em> 4 (2000), 451&ndash;466.</p>
<p><a name="R10" id="R10"></a>10. Hindley, R. The principal type-scheme of an object in combinatory logic. <em>Trans. Am. Math. Soc.</em> (1969), 29&ndash;60.</p>
<p><a name="R11" id="R11"></a>11. Jump, M., McKinley, K.S. Cork: Dynamic memory leak detection for garbage-collected languages. In <em>ACM SIGPLAN Notices</em>, Volume 42 (2007). ACM, 31&ndash;38.</p>
<p><a name="R12" id="R12"></a>12. Kleinschmager, S., Hanenberg, S., Robbes, R., Tanter, &Eacute;., Stefik, A. Do static type systems improve the maintainability of software systems? An empirical study. In <em>2012 IEEE 20th International Conference on Program Comprehension (ICPC)</em> (2012). IEEE, 153&ndash;162.</p>
<p><a name="R13" id="R13"></a>13. Li, Z., Tan, L., Wang, X., Lu, S., Zhou, Y., Zhai, C. Have things changed now? An empirical study of bug characteristics in modern open source software. In <em>ASID'06: Proceedings of the 1st Workshop on Architectural and System Support for Improving Software Dependability</em> (October 2006).</p>
<p><a name="R14" id="R14"></a>14. Marques De S&aacute;, J.P. <em>Applied Statistics Using SPSS, Statistica and Matlab</em>, 2003.</p>
<p><a name="R15" id="R15"></a>15. Mayer, C., Hanenberg, S., Robbes, R., Tanter, &Eacute;., Stefik, A. An empirical study of the influence of static type systems on the usability of undocumented software. In <em>ACM SIGPLAN Notices</em>, Volume 47 (2012). ACM, 683&ndash;702.</p>
<p><a name="R16" id="R16"></a>16. Meyerovich, L.A., Rabkin, A.S. Empirical analysis of programming language adoption. In <em>Proceedings of the 2013 ACM SIGPLAN International Conference on Object Oriented Programming Systems Languages &amp; Applications</em> (2013). ACM, 1&ndash;18.</p>
<p><a name="R17" id="R17"></a>17. Milner, R. A theory of type polymorphism in programming. <em>J. Comput. Syst. Sci. 17</em>, 3 (1978), 348&ndash;375.</p>
<p><a name="R18" id="R18"></a>18. Mockus, A., Votta, L.G. Identifying reasons for software changes using historic databases. In <em>ICSM'00. Proceedings of the International Conference on Software Maintenance</em> (2000). IEEE Computer Society, 120.</p>
<p><a name="R19" id="R19"></a>19. Odersky, M., Spoon, L., Venners, B. <em>Programming in Scala.</em> Artima Inc, 2008.</p>
<p><a name="R20" id="R20"></a>20. Pankratius, V., Schmidt, F., Garret&oacute;n, G. Combining functional and imperative programming for multicore software: An empirical study evaluating scala and java. In <em>Proceedings of the 2012 International Conference on Software Engineering</em> (2012). IEEE Press, 123&ndash;133.</p>
<p><a name="R21" id="R21"></a>21. Petricek, T., Skeet, J. <em>Real World Functional Programming: With Examples in F# and C#.</em> Manning Publications Co., 2009.</p>
<p><a name="R22" id="R22"></a>22. Pierce, B.C. <em>Types and Programming Languages.</em> MIT Press, 2002.</p>
<p><a name="R23" id="R23"></a>23. Posnett, D., Bird, C., D&eacute;vanbu, P. An empirical study on the influence of pattern roles on change-proneness. <em>Emp. Softw. Eng. 16</em>, 3 (2011), 396&ndash;423.</p>
<p><a name="R24" id="R24"></a>24. Tan, L., Liu, C., Li, Z., Wang, X., Zhou, Y., Zhai, C. Bug characteristics in open source software. <em>Emp. Softw. Eng.</em> (2013).</p>
</div>
<p><a href="#PageTop">Back to Top</a></p>
<div>
<a name="authorinfo" id="authorinfo"></a>
<h3>Authors</h3>
<p><strong>Baishakhi Ray</strong> (<a href="http://delivery.acm.org/10.1145/3130000/3126905/mailto:rayb@virginia.edu">rayb@virginia.edu</a>), Department of Computer Science, University of Virginia, Charlottesville, VA.</p>
<p><strong>Daryl Posnett</strong> (<a href="http://delivery.acm.org/10.1145/3130000/3126905/mailto:dpposnett@ucdavis.edu">dpposnett@ucdavis.edu</a>), Department of Computer Science, University of California, Davis, CA.</p>
<p><strong>Premkumar Devanbu</strong> (<a href="http://delivery.acm.org/10.1145/3130000/3126905/mailto:devanbu@cs.ucdavis.edu">devanbu@cs.ucdavis.edu</a>), Department of Computer Science, University of California, Davis, CA.</p>
<p><strong>Vladimir Filkov</strong> (<a href="http://delivery.acm.org/10.1145/3130000/3126905/mailto:filkov@cs.ucdavis.edu">filkov@cs.ucdavis.edu</a>), Department of Computer Science, University of California, Davis, CA.</p>
</div>
<p><a href="#PageTop">Back to Top</a></p>
<div>
<hr>
<a name="permission" id="permission"></a>
<p>Copyright held by owners/authors. Publication rights licensed to ACM.<br>
Request permission to publish from <a href="http://delivery.acm.org/10.1145/3130000/3126905/mailto:permissions@acm.org">permissions@acm.org</a></p>
</div>
<p>The Digital Library is published by the Association for Computing Machinery. Copyright&nbsp;&copy;&nbsp;2017 ACM, Inc.</p>
<hr>
<a name="comments" id="comments"></a>
<p>No entries found</p>
</div>
<a href="https://cacm.acm.org/magazines/2017/10/221326-a-large-scale-study-of-programming-languages-and-code-quality-in-github/"></a></section>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via cacm.acm.org https://cacm.acm.org<br>
<br>
May 2, 2020 at 11:41AM
